From 3113b16d28479e355bfc18478cf96956b99969a7 Mon Sep 17 00:00:00 2001
From: q66 <q66@chimera-linux.org>
Date: Wed, 13 Dec 2023 05:14:28 +0100
Subject: [PATCH] timeout(1): rudimentary fix for non-foreground behavior

This is probably still not robust but is better than the previous
broken behavior carried over from bsdutils.
---
 patches/src.freebsd.patch               | 16 +++++++---------
 src.freebsd/coreutils/timeout/timeout.c |  4 +---
 2 files changed, 8 insertions(+), 12 deletions(-)

diff --git a/src.freebsd/coreutils/timeout/timeout.c b/src.freebsd/coreutils/timeout/timeout.c
index ea1d494..e882d13 100644
--- a/src.freebsd/coreutils/timeout/timeout.c
+++ b/src.freebsd/coreutils/timeout/timeout.c
@@ -175,7 +175,6 @@ main(int argc, char **argv)
 	bool do_second_kill = false;
 	bool child_done = false;
 	struct sigaction signals;
-	unsigned long info;
 	int signums[] = {
 		-1,
 		SIGTERM,
@@ -297,8 +296,7 @@ main(int argc, char **argv)
 				if (foreground) {
 					break;
 				} else {
-					prctl(PR_GET_CHILD_SUBREAPER, &info, 0, 0);
-					if (info == 0)
+					if (waitpid(-1, NULL, WNOHANG) < 0)
 						break;
 				}
 			}
